pyro/contrib/epidemiology/sir.py

```diff
@@ -318,3 +322,153 @@ def transition_bwd(self, params, prev, curr, t):
+    3.  How to override the :meth:`predict` method to compute extra
+        statistics.
+
+    :param int population: Total ``population = S + I + R``.
+    :param float recovery_time: Mean recovery time (duration in state
+        ``I``). Must be greater than 1.
+    :param int pre_window: Number of time steps before beginning ``data``
+        where the initial infection may have occurred. Must be positive.
+    :param iterable data: Time series of new observed infections. Each time
+        step is Binomial distributed between 0 and the number of ``S -> I``
+        transitions. This allows false negative but no false positives.
+    """
+
+    def __init__(self, population, recovery_time, pre_window, data):
+        compartments = ("S", "I")  # R is implicit.
+        duration = pre_window + len(data)
+        super().__init__(compartments, duration, population)
+
+        assert isinstance(recovery_time, float)
+        assert recovery_time > 1
+        self.recovery_time = recovery_time
+
+        assert isinstance(pre_window, int) and pre_window > 0
+        self.pre_window = pre_window
+        self.post_window = len(data)
+
+        # Expect a single external infection during the pre_window.
+        self.external_rate = 1 / pre_window
+
+        # Prepend data with zeros.
+        if isinstance(data, list):
+            data = [0.] * self.pre_window + data
+        else:
+            data = pad(data, (self.pre_window, 0), value=0.)
+        self.data = data
+
+    series = ("S2I", "I2R", "obs")
+    full_mass = [("R0", "rho0", "rho1")]
+
+    def global_model(self):
+        tau = self.recovery_time
+        R0 = pyro.sample("R0", dist.LogNormal(0., 1.))
+
+        # Assume two different response rates: rho0 before any observations
+        # were made (in pre_window), followed by a higher response rate rho1
+        # after observations were made (in post_window).
+        rho0 = pyro.sample("rho0", dist.Uniform(0, 1))
+        rho1 = pyro.sample("rho1", dist.Uniform(0, 1))
+        rho = torch.cat([
+            rho0.unsqueeze(-1).expand(rho0.shape + (self.pre_window,)),
+            rho1.unsqueeze(-1).expand(rho1.shape + (self.post_window,)),
+        ], dim=-1)
+
+        # Model external infections as an infectious pseudo-individual added
+        # to num_infectious when sampling S2I below.
+        X = self.external_rate * tau / R0
+
+        return R0, X, tau, rho
+
+    def initialize(self, params):
+        # Start with no internal infections.
+        return {"S": self.population, "I": 0}
+
+    def transition_fwd(self, params, state, t):
+        R0, X, tau, rho = params
+
+        # Sample flows between compartments.
+        S2I = pyro.sample("S2I_{}".format(t),
+                          infection_dist(individual_rate=R0 / tau,
+                                         num_susceptible=state["S"],
+                                         num_infectious=state["I"] + X,
```

Comment:
i guess the `X` contribution is active at all times?

Reply:
Correct. That modeling decision seems more parsimonious, simpler to code, and quantitatively close to the alternative (one initial infection at an unknown time) because during an outbreak, new infections should be dominated by local internal infections.